trin-core/tests/overlay.rs

```diff
@@ -0,0 +1,223 @@
+    let overlay_config = OverlayConfig::default();
+    let utp_listener = Arc::new(RwLock::new(UtpListener {
+        discovery: Arc::clone(&discovery),
+        utp_connections: HashMap::new(),
+        listening: HashMap::new(),
+    }));
+
+    OverlayProtocol::new(
+        overlay_config,
+        discovery,
+        utp_listener,
+        db,
+        U256::MAX,
+        protocol,
+    )
+    .await
+}
+
+async fn spawn_overlay(discovery: Arc<Discovery>, overlay: Arc<OverlayProtocol>) {
+    let (overlay_tx, mut overlay_rx) = mpsc::unbounded_channel();
+    let mut discovery_rx = discovery
+        .discv5
+        .event_stream()
+        .await
+        .map_err(|err| err.to_string())
+        .unwrap();
+
+    let overlay_protocol = overlay.protocol().clone();
+    tokio::spawn(async move {
+        while let Some(discovery_event) = discovery_rx.recv().await {
+            let talk_req = match discovery_event {
+                Discv5Event::TalkRequest(req) => req,
+                _ => continue,
+            };
+
+            let req_protocol = ProtocolId::from_str(&hex::encode_upper(talk_req.protocol()));
+
+            if let Ok(req_protocol) = req_protocol {
+                match (req_protocol, overlay_protocol.clone()) {
+                    (ProtocolId::History, ProtocolId::History)
+                    | (ProtocolId::State, ProtocolId::State) => overlay_tx.send(talk_req).unwrap(),
+                    _ => panic!("Unexpected protocol"),
+                }
+            } else {
+                panic!("Invalid protocol");
+            }
+        }
+    });
+
+    tokio::spawn(async move {
+        while let Some(talk_req) = overlay_rx.recv().await {
+            let talk_resp = match overlay.process_one_request(&talk_req).await {
+                Ok(response) => Message::Response(response).to_bytes(),
+                Err(err) => panic!("Error processing request: {}", err),
+            };
+            if let Err(err) = talk_req.respond(talk_resp) {
+                panic!("Unable to respond to talk request: {}", err);
+            }
+        }
+    });
+}
+
+// Basic tests for overlay routing table management according to messages exchanged between
+// multiple nodes.
+//
+// Use sleeps to give time for background routing table processes.
+#[tokio::test]
+async fn overlay() {
+    let protocol = ProtocolId::History;
+
+    // Node one.
+    let portal_config_one = PortalnetConfig {
+        listen_port: 8001,
+        internal_ip: true,
+        ..PortalnetConfig::default()
+    };
+    let mut discovery_one = Discovery::new(portal_config_one).unwrap();
+    let _ = discovery_one.start().await.unwrap();
+    let discovery_one = Arc::new(discovery_one);
+    let overlay_one = Arc::new(init_overlay(Arc::clone(&discovery_one), protocol.clone()).await);
+    spawn_overlay(Arc::clone(&discovery_one), Arc::clone(&overlay_one)).await;
+    time::sleep(Duration::from_millis(5)).await;
```

Comment:
nitpick: It may be useful to extract the milliseconds' value in a constant above the tests.